nemo-physical/src/dictionary/ranked_pair_dictionary.rs

```diff
@@ -0,0 +1,499 @@
+//! This module defines dictionaries for pairs of byte arrays (resp. strings) where
+//! one part is "frequent" (few values used often) and another is "rare"
+//! (many values, used rarely).
+
+use crate::management::bytesized::ByteSized;
+
+use super::{
+    bytes_buffer::{BytesBuffer, GlobalBytesBuffer},
+    bytes_dictionary::BytesDictionary,
+    AddResult,
+};
+
+/// Smallest "frequent ID" that cannot be stored in a single byte.
+///
+/// We store IDs of frequent parts of pairs using 1-4 bytes, where
+/// the first two bits indicate how many bytes were used.
+const MAX_ID_1B: usize = 1 << 6;
+/// Smallest "frequent ID" that cannot be stored in two bytes.
+/// See [MAX_ID_1B].
+const MAX_ID_2B: usize = 1 << (6 + 8);
+/// Smallest "frequent ID" that cannot be stored in three bytes.
+/// See [MAX_ID_1B].
+const MAX_ID_3B: usize = 1 << (6 + 8 + 8);
+/// Smallest "frequent ID" that cannot be stored in four bytes.
+/// See [MAX_ID_1B].
+const MAX_ID_4B: usize = 1 << (6 + 8 + 8 + 8);
+
+/// Number of bytes in usize. Just for making the code below more readable.
+const USIZE_BYTES: usize = size_of::<usize>();
+/// The bits we consider for encoding information about the length of the frequent ids
+/// in our encoding.
+const FTYPE_BIT_MASK: u8 = 0b1100_0000;
+/// The bits in the first frequent id byte that are part of the id, without the bits
+/// used to encode the type.
+const FID_BIT_MASK: u8 = 0b0011_1111;
+
+/// Simple least-recently used data structure that can be used as a cache for
+/// mappings from byte arrays to usize ids. The implementation is intended for
+/// very small cache sizes, where the most efficient way of getting an item is
+/// to iterate over all entries.
+#[derive(Debug)]
+struct LruArray {
+    cache: Vec<(Vec<u8>, usize)>,
+    top: usize,
+    len: usize,
+}
+impl LruArray {
+    fn create(capacity: usize) -> Self {
```

Comment:
The constructor should be called `new`, cf. the [Rust API Guidelines](https://rust-lang.github.io/api-guidelines/predictability.html). Also, `capacity` should be `std::num::NonZero<usize>` to make this sound.